{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules\n",
    "import pandas as pd # v0.23.4\n",
    "from matplotlib import pyplot as plt # v3.0.2\n",
    "import matplotlib.gridspec as gridspec # v3.0.2\n",
    "from scipy import stats # v1.1.0\n",
    "import numpy as np # v1.15.4\n",
    "import itertools\n",
    "import seaborn as sns # v0.9.0\n",
    "from Bio import SeqIO # v1.73"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import table containing the PCR marker results and sporulation ability\n",
    "with open(\"mito_markers.csv\",\"r\") as file_in:\n",
    "    data=pd.read_csv(file_in, header=0)\n",
    "NC=dict(zip([\"A\",\"B\",\"C\",\"D\",\"E\",\"F\",\"H\",\"I\"],[\"L1\",\"M1\",\"H1\",\"L2\",\"M2\",\"H2\",\"VL2\",\"VL1\"]))\n",
    "Time=dict(zip([\"Tinit\",\"Tfin\"],[r\"$T_{ini}$\",r\"$T_{end}$\"]))\n",
    "mito_genotype=[]\n",
    "nuc_genotype=[]\n",
    "for i in data.index:\n",
    "    if data.loc[i,\"rnl\"]==\"+\" and data.loc[i,\"atp6\"]==\"+\":\n",
    "        # If both rnl and atp6 markers are present, score mitochondrial completeness as \"+\"\n",
    "        mito_genotype.append(\"+\")\n",
    "    else:\n",
    "        # If any of rnl or atp6 markers is absent, score mitochondrial completeness as \"-\"\n",
    "        mito_genotype.append(\"-\")\n",
    "    nuc_genotype.append(NC[data.loc[i,\"strain\"][0]])\n",
    "data[\"mtg\"]=mito_genotype\n",
    "data[\"ncg\"]=nuc_genotype\n",
    "# Add formatted time values\n",
    "DATA=data.copy()\n",
    "DATA[\"time\"]=[Time[i] for i in DATA[\"time\"]]\n",
    "\n",
    "# Create contingency tables for Tend and Tini\n",
    "contingency1=pd.DataFrame(np.zeros([2,2]), index=[\"spo+\",\"spo-\"], columns=[\"mito+\", \"mito-\"])\n",
    "contingency2=pd.DataFrame(np.zeros([2,2]), index=[\"spo+\",\"spo-\"], columns=[\"mito+\", \"mito-\"])\n",
    "for i in data.index:\n",
    "    for j,k in itertools.product([\"+\",\"-\"],[\"+\",\"-\"]):\n",
    "        if data.loc[i,\"mtg\"]==j and data.loc[i,\"sporulation\"]==k:\n",
    "            if data.loc[i,\"time\"]==\"Tfin\":\n",
    "                contingency1.loc[\"spo\"+k, \"mito\"+j]+=1\n",
    "            if data.loc[i,\"time\"]==\"Tinit\":\n",
    "                contingency2.loc[\"spo\"+k, \"mito\"+j]+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Export the mito markers data for figure creation\n",
    "\n",
    "fig, axes = plt.subplots(ncols=3, nrows=4, figsize=[12,12])\n",
    "ax_idx=0\n",
    "for i in [range(32), range(32,64), range(64,80)]:\n",
    "\n",
    "    plot_table=np.array(DATA.loc[i, [\"ncg\",\"strain\",\"time\",\"sporulation\"]])\n",
    "    ax=axes[3,ax_idx]\n",
    "    table=ax.table(cellText=plot_table, loc=\"top\",\n",
    "             cellLoc = 'center', rowLoc = 'center')\n",
    "\n",
    "    for cell in table.properties()[\"child_artists\"]:\n",
    "        cell.set_height(0.09)\n",
    "        cell.set_width(0.20)\n",
    "    ax_idx=ax_idx+1\n",
    "    \n",
    "for i,j in itertools.product(range(4), range(3)):\n",
    "    axes[i,j].axis(\"off\")\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"mito_table.pdf\")\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Plot the contingency tables\n",
    "fig = plt.figure(figsize=[6.5,3.2])\n",
    "\n",
    "GS=gridspec.GridSpec(nrows=1, ncols=2, width_ratios=[1,1.25])\n",
    "\n",
    "ax = fig.add_subplot(GS[0,0])\n",
    "\n",
    "sns.heatmap(contingency2, annot=True, cmap=\"Blues\", ax=ax,\n",
    "            cbar=False,\n",
    "           vmin=0, vmax=40, linecolor=\"black\", linewidths=1)\n",
    "ax.set_xlabel(\"mtDNA completeness\", size=14)\n",
    "ax.set_xticklabels([\"+\",\"-\"], size=14)\n",
    "ax.set_ylabel(\"Sporulation ability\", size=14)\n",
    "ax.set_yticklabels([\"+\",\"-\"], rotation=0, size=14)\n",
    "ax.set_title(r\"$T_{ini}$\", size=18)\n",
    "ax.xaxis.set_ticks_position('none')\n",
    "ax.yaxis.set_ticks_position('none')\n",
    "\n",
    "ax = fig.add_subplot(GS[0,1])\n",
    "HM=sns.heatmap(contingency1, annot=True, cmap=\"Blues\", ax=ax,\n",
    "            cbar=True, cbar_kws={\"label\":\"Number of lines\"},\n",
    "           vmin=0, vmax=40, linecolor=\"black\", linewidths=1)\n",
    "ax.set_xlabel(\"mtDNA completeness\", size=14)\n",
    "ax.set_xticklabels([\"+\",\"-\"], size=14)\n",
    "ax.set_ylabel(\"Sporulation ability\", size=14)\n",
    "ax.set_yticklabels([\"+\",\"-\"], rotation=0, size=14)\n",
    "ax.set_title(r\"$T_{end}$\", size=18)\n",
    "ax.xaxis.set_ticks_position('none')\n",
    "ax.yaxis.set_ticks_position('none')\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"mtDNAcomplete_sporulation.pdf\")\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Make Fisher's exact test on the data at Tend, replacing the 0 value for [spo+, mito-] by one to \n",
    "# yield a finite boundary to the odds ratio.\n",
    "test=contingency1.copy()\n",
    "test.loc[\"spo+\",\"mito-\"]=1\n",
    "stats.fisher_exact(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Plot a map of the mitochondrial genomes and annotations with positions of the PCR markers\n",
    "# import mito genome and annotations from S288C and YPS128, from https://yjx1217.github.io/Yeast_PacBio_2016/data/\n",
    "strains=zip([\"S288c.mt\",\n",
    "        \"YPS138.mt\"],\n",
    "           [[(192,477), (31379,32168)], [(192,477), (25969,26687)]])\n",
    "\n",
    "fig, axes = plt.subplots(ncols=2, figsize=[10,5], subplot_kw={\"projection\":\"polar\"})\n",
    "lwd=1\n",
    "\n",
    "idx=0\n",
    "\n",
    "for s, loc in strains:\n",
    "    strain_name=s.split(\"/\")[-1][:-3]\n",
    "    with open(s+\".genome.fa\",\"r\") as mito_in:\n",
    "        mtgenome=[i for i in SeqIO.parse(mito_in, \"fasta\")]\n",
    "    with open(s+\".all_feature.gff\",\"r\") as gff_in:\n",
    "        mtgff=pd.DataFrame([i.split(\"\\t\") for i in gff_in.read().splitlines()])\n",
    "    mtgff[3]=mtgff[3].astype(int)\n",
    "    mtgff[4]=mtgff[4].astype(int)\n",
    "    NAME=[]\n",
    "    ID=[]\n",
    "    for i in mtgff.index:\n",
    "        attributes=dict([(j.split(\"=\")[0],j.split(\"=\")[1]) for j in mtgff.loc[i,8].split(\";\")])\n",
    "        NAME.append(attributes[\"Name\"])\n",
    "        ID.append(attributes[\"ID\"])\n",
    "    mtgff[\"ID\"]=ID\n",
    "    mtgff[\"Name\"]=NAME\n",
    "\n",
    "    # plot a schematic map of mitochondrial genome\n",
    "    len_genome=len(mtgenome[0].seq)\n",
    "    def circ_coord(x):\n",
    "        c=2*np.pi*x/len_genome\n",
    "        return(c)\n",
    "\n",
    "    genome_backbone=np.arange(0,circ_coord(len_genome),0.01)\n",
    "    genes_mito=mtgff.loc[mtgff[2]!=\"mRNA\"]\n",
    "\n",
    "    ax=axes[idx]\n",
    "\n",
    "    ax.set_theta_zero_location(\"N\")\n",
    "    ax.set_theta_direction(\"clockwise\")\n",
    "\n",
    "    ax.plot(genome_backbone, np.repeat(4,len(genome_backbone)), linewidth=1, color=\"black\")\n",
    "\n",
    "    for i in genes_mito.index:\n",
    "\n",
    "        genespan=np.arange(circ_coord(mtgff.loc[i,3]-1),circ_coord(mtgff.loc[i,4]),0.001)\n",
    "        midpos=0.5*sum([circ_coord(mtgff.loc[i,3]), circ_coord(mtgff.loc[i,4])])\n",
    "\n",
    "        if mtgff.loc[i,2]==\"gene\":\n",
    "            col=\"cyan\"\n",
    "            #lwd=4\n",
    "            name=mtgff.loc[i,\"Name\"]\n",
    "\n",
    "        if mtgff.loc[i,2]==\"tRNA\":\n",
    "            col=\"lime\"\n",
    "            #lwd=4\n",
    "            name=None\n",
    "        if mtgff.loc[i,2]==\"CDS\":\n",
    "            col=\"blue\"\n",
    "            #lwd=4\n",
    "            name=None\n",
    "        if mtgff.loc[i,2] in ['rnl', 'rnpB', 'rns']:\n",
    "            col=\"red\"\n",
    "            #lwd=4\n",
    "            name=mtgff.loc[i,2]\n",
    "\n",
    "        for y in np.arange(3.8,4.21,0.03):\n",
    "            ax.plot(genespan, np.repeat(y,len(genespan)), linewidth=lwd, color=col)\n",
    "\n",
    "        # if a protein or RNA-coding gene, print tag\n",
    "        if mtgff.loc[i,2] in [\"gene\", 'rnl', 'rnpB', 'rns']:\n",
    "\n",
    "\n",
    "            ax.plot([midpos, midpos], [4.35,4.55], linewidth=0.5, color=\"black\")\n",
    "\n",
    "            if 0<=midpos<=np.pi or 2*np.pi-0.2<=midpos<2*np.pi:\n",
    "                ax.text(midpos, 5.3, s=name, ha=\"center\", va=\"center\", \n",
    "                        rotation=90-180*(midpos/np.pi),\n",
    "                       size=11, style=\"italic\") \n",
    "\n",
    "            if np.pi<=midpos<2*np.pi-0.2:\n",
    "                ax.text(midpos, 5.3, s=name, ha=\"center\", va=\"center\", \n",
    "                       rotation=-90-180*(midpos)/np.pi,\n",
    "                       size=11, style=\"italic\")\n",
    "\n",
    "    for i in loc:\n",
    "\n",
    "        genespan_pcr=np.arange(circ_coord(i[0]),circ_coord(i[1]),0.001)\n",
    "        midpos=0.5*sum([circ_coord(i[0]),circ_coord(i[1])])\n",
    "        for y in np.arange(3.3,3.71,0.03):\n",
    "            ax.plot(genespan_pcr, np.repeat(y,len(genespan_pcr)), linewidth=lwd, color=\"orange\")\n",
    "\n",
    "    for i in np.arange(0,len_genome,10000):\n",
    "        ax.scatter(circ_coord(i),3, marker=\".\", color=\"black\", s=5)\n",
    "        label=\"%.0f\" % (i/1000)\n",
    "        ha=\"center\"\n",
    "        if i==0:\n",
    "            label=\"%.0f kb\" % (i/1000)\n",
    "            if strain_name==\"S288c\":\n",
    "                ha=\"center\"\n",
    "            if strain_name==\"YPS138\":\n",
    "                ha=\"left\"\n",
    "        if i==70000 and strain_name==\"YPS138\":\n",
    "            label=\"%.0f\" % (i/1000)\n",
    "            ha=\"right\"            \n",
    "\n",
    "        ax.text(circ_coord(i), 2.5, s=label,\n",
    "               ha=ha, va=\"center\", size=10)\n",
    "    ax.text(0,0, s=strain_name, size=18, ha=\"center\", va=\"center\")\n",
    "\n",
    "    ax.grid(False)\n",
    "    ax.axis(\"off\")\n",
    "\n",
    "\n",
    "    ax.set_ylim([0,5.5])\n",
    "    idx=idx+1\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"mito_map_5x10.pdf\")\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
